puffin/src/trace.rs

```diff
@@ -110,6 +112,32 @@ pub struct TraceContext<PB: ProtocolBehavior + 'static> {
     phantom: PhantomData<PB>,
 }
 
+impl<PB: ProtocolBehavior + 'static> fmt::Display for TraceContext<PB> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "Knowledge [not displaying other fields] (size={}):",
+            self.knowledge.len()
+        );
+        for k in &self.knowledge {
+            write!(f, "\n   {},          --  {:?}", k, k);
+        }
+        Ok(())
+    }
+}
+
+impl<PB: ProtocolBehavior + PartialEq> PartialEq for TraceContext<PB> {
+    fn eq(&self, other: &Self) -> bool {
+        self.agents == other.agents
+            && self.put_registry == other.put_registry
+            && self.deterministic_put == other.deterministic_put
+            && self.default_put_options == other.default_put_options
+            && self.non_default_put_descriptors == other.non_default_put_descriptors
+            && format!("{:?}", self.knowledge) == format!("{:?}", other.knowledge)
+            && format!("{:?}", self.claims) == format!("{:?}", other.claims)
```

Comment:
This seems a bit hacky. Can we derive PartialEq?

Reply:
We do not impose equality of all fields. We could have used the [derivative](https://crates.io/crates/derivative) to exclude those but I didn't know this crate at the time. Plus, this is solely used for testing the determinism.